Added AsyncManualResetEvent

noescape00 opened the pull request:
AsyncManualResetEvent is an async version of ManualResetEventSlim

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,269 @@
+        /// Waits for the task to complete, unwrapping any exceptions.
+        /// </summary>
+        /// <param name="task">The task. May not be <c>null</c>.</param>
+        private static void WaitAndUnwrapException(Task task)
```
we don't need this method, nor its other version below

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,269 @@
+            if (!cancellationToken.CanBeCanceled)
+                return task;
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled(cancellationToken);
```
after each return there should be empty space everywhere if the code continues - i.e. there is no `}` following

noescape00 replied:
Fixed

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,269 @@
+        private static void WaitAndUnwrapException(Task task)
+        {
+            if (task == null)
+                throw new ArgumentNullException(nameof(task));
```
also after throw make a new empty line everywhere

noescape00 replied:
Fixed

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+            return WaitAsync(waitTask, cancellationToken);
+        }
+
+        /// <summary>
```
i think we don't need this method,
 if we want to wait synchrounously, we will do what we do elsewhere - .GetAwaiter().GetResult()

noescape00 replied:
Ok. Wanted to keep it so we can use AsyncManualResetEvent as a normal ManualResetEventSlim but you're right, most likely this methods will never be used.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        /// Synchronously waits for this event to be set. This method may block the calling thread.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token used to cancel the wait. If this token is already canceled, this method will first check whether the event is set.</param>
+        public void Wait(CancellationToken cancellationToken)
```
also not needed imho

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        /// Waits for the task to complete, unwrapping any exceptions.
+        /// </summary>
+        /// <param name="task">The task. May not be <c>null</c>.</param>
+        private static void WaitAndUnwrapException(Task task)
```
since i think we don't need sync Wait() methods we can delete this method

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        /// <param name="task">The task. May not be <c>null</c>.</param>
+        /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
+        /// <exception cref="OperationCanceledException">The <paramref name="cancellationToken"/> was cancelled before the <paramref name="task"/> completed, or the <paramref name="task"/> raised an <see cref="OperationCanceledException"/>.</exception>
+        private static void WaitAndUnwrapException(Task task, CancellationToken cancellationToken)
```
synce i think we don't need Wait() method, we can delete this one

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        /// </summary>
+        /// <param name="exception">The exception. May not be <c>null</c>.</param>
+        /// <returns>The <see cref="Exception"/> that was passed into this method.</returns>
+        private static Exception PrepareForRethrow(Exception exception)
```
and we can delete this one if we delete that one

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+            return DoWaitAsync(task, cancellationToken);
+        }
+
+        private static async Task DoWaitAsync(Task task, CancellationToken cancellationToken)
```
we can get rid of this method if we just inline its code into the above WaitAsync method

if not, we should document it

noescape00 replied:
Documented.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        /// <param name="cancellationToken">The cancellation token that cancels the wait.</param>
+        private static Task WaitAsync(Task task, CancellationToken cancellationToken)
+        {
+            if (task == null)
```
we will use Guards in our codebase for this, pls check whole code

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        private static async Task DoWaitAsync(Task task, CancellationToken cancellationToken)
+        {
+            using (var cancelTaskSource = new CancellationTokenTaskSource<object>(cancellationToken))
+                await await Task.WhenAny(task, cancelTaskSource.Task).ConfigureAwait(false);
```
the double await construct probably deserves to be commented,
it is explained here http://www.tomdupont.net/2014/09/await-await-taskwhenany.html

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+
+        private static async Task DoWaitAsync(Task task, CancellationToken cancellationToken)
+        {
+            using (var cancelTaskSource = new CancellationTokenTaskSource<object>(cancellationToken))
```
the whole CancellationTokenTaskSource class seems unnecessary to me, it has one simply use that can be inlined imho

so basically we can do using over the registration instead of using over this

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,273 @@
+        /// Creates a new TCS for use with async code, and which forces its continuations to execute asynchronously.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result of the TCS.</typeparam>
+        private static TaskCompletionSource<TResult> CreateAsyncTaskSource<TResult>()
```
i also think we want to convert all these static methods to nonstatic methods

noescape00 replied:
Fixed.

noescape00 commented:
>i'd do some more changes, rip off some methods that we don't need
I'm not sure if we can go further. The class is already pretty stripped, even the default even slim's Wait() is deleted in order to keep it shorter.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,177 @@
+        /// <summary>
+        /// Waits for the task to be executed or canceled.
+        /// </summary>
+        /// <remarks>
```
remarks should be under param section

noescape00 replied:
Fixed

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,177 @@
+        /// </summary>
+        /// <remarks>
+        /// Double await construct is used here in order to catch exceptions in child tasks.
+        /// It is similar to await + unwrap but uses less allocations. <see cref="https://stackoverflow.com/questions/34816628/await-await-vs-unwrap"/>
```
if you use `<see>` you need to put it into a sentence, or use `seealso` section

so for example

`For more information about double await see <see ...>.`

noescape00 replied:
Fixed

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,177 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+https://github.com/StephenCleary/AsyncEx
```
here I would change this line to:

```
Based on the original implementation of AsyncManualResetEventfrom Stephen Cleary's AsyncEx library - https://github.com/StephenCleary/AsyncEx
```

noescape00 replied:
Fixed.

noescape00 commented:
>however, as we changed the original implementation quite a lot, i think we absolutely need tests that will test all functionality of this object

Added unit tests.

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,98 @@
+    public class AsyncManualResetEventTest
+    {
+        [Fact]
+        public async void AsyncManualResetEvent_WaitAsync()
```
what is the purpose of this test? (there is no comment ...)
it seems to test that
- using AMRE does not crash
- WaitAsync eventually completes

however, it does not check
- whether WaitAsync completes when the event is set

also please use lambdas over delegates to be consistent with the rest of the codebase

noescape00 replied:
Fixed.

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        [Fact]
+        public void AsyncManualResetEvent_CanReset()
```
we could have CanSetAndReset and test setting the event again and checking ...

noescape00 replied:
Fixed.

noescape00 commented:
>i'm also missing tests with cancellation tokens - so please add those

>and one more thing - once we removed static from many methods, we need to add this. when we call them

Fixed.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,179 @@
+
+            var tcs = new TaskCompletionSource<object>();
+            using (cancellationToken.Register(() => tcs.TrySetCanceled(cancellationToken), useSynchronizationContext: false))
+                await await Task.WhenAny(task, tcs.Task).ConfigureAwait(false);
```
i believe we do want `{}` around `lock` and `using` statements

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,179 @@
+    public sealed class AsyncManualResetEvent
+    {
+        /// <summary>
+        /// The object used for synchronization.
```
this is not good enough comment (together with the comment of tcs), see e.g. block puller locks

what you want here is to describe the contract fully for both the lock and the protected object
so here 
`Lock to protect access to <see cref="tcs"/>.`

and in tcs, you should add
```
<remarks>All access to this object has to be protected by <see cref="mutex"/>.</remarks>
```

only then the reader could easily understand that the mutex is there to protect tcs and then when tcs is touched it should be protected.

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,179 @@
+            if (waitTask.IsCompleted)
+                return waitTask;
+
+            return this.WaitAsync(waitTask, cancellationToken);
```
this call does not make much sense
the body of that method should be copied here and that method should be removed
otherwise you would have that method inside this class that does not work with this class instance at all (remember it was static before -> so that's why it should be brought here)

Aprogiena replied:
and then you will find you don't need DoWaitAsync either

noescape00 replied:
I'd actually keep the DoWaitAsync and not merge it with WaitAsync because if we merge it we should make the code that is now in DoWaitAsync to be synchronous or change the signature of WaitAsync which I'd avoid.

noescape00 replied:
>the body of that method should be copied here and that method should be removed

Ok.

Aprogiena replied:
but is it not worth it to remove DWA?

Aprogiena replied:
how about something like this, to me it looks simple enough for it to be 1 method, not 3
it remains to check if i did not do any mistakes with my quick hack coding 
```
        public async Task WaitAsync(CancellationToken cancellationToken)
        {
            Task waitTask = this.WaitAsync();
            if (waitTask.IsCompleted)
                return;

            if (!cancellationToken.CanBeCanceled)
            {
                await waitTask.ConfigureAwait(false);
                return;
            }

            cancellationToken.ThrowIfCancellationRequested();

            var tcs = new TaskCompletionSource<object>();
            using (cancellationToken.Register(() => tcs.TrySetCanceled(cancellationToken), useSynchronizationContext: false))
            {
                await await Task.WhenAny(waitTask, tcs.Task).ConfigureAwait(false);
            }
        }
```

so if you could check this is really equivalent of the original ... then let me know if you don't think this is better

Aprogiena commented on src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs:
```diff
@@ -0,0 +1,179 @@
+        }
+
+        /// <summary>
+        /// Sets the event, atomically completing every task returned by <see cref="O:Nito.AsyncEx.AsyncManualResetEvent.WaitAsync"/>. If the event is already set, this method does nothing.
```
certainly not O:Nito.AsyncEx ....

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        [Fact]
+        public async void AsyncManualResetEvent_CanBeCancelled()
```
Async suffix

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+
+            mre.Set();
+            mre.Reset();
+            var task = mre.WaitAsync();
```
unvar these everywhere

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+        /// <summary>
+        /// Tests that:
+        /// <list type="bullet">
+        /// <item><see cref="AsyncManualResetEvent"/> does not crush.</item>
```
well :-) that was kind of a joke to put in the contrast what it really tests and what it does not test
to point out that maybe it was intentional to actually test what it did not test ...

noescape00 replied:
Kk, I'll remove the comment.

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+        /// Tests that:
+        /// <list type="bullet">
+        /// <item><see cref="AsyncManualResetEvent"/> does not crush.</item>
+        /// <item>WaitAsync eventually completes.</item>
```
always `<see>` when you refer to a method or class or whatever

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+
+            stopwatch.Stop();
+
+            Assert.True(stopwatch.ElapsedMilliseconds >= 500 && stopwatch.ElapsedMilliseconds < 600);
```
this could blow as we've seen in my stopwatch tests
although you did allow 100 ms epsilon, so most of the time it should be ok, but better to avoid this
also use () or make two separate Asserts

Aprogiena replied:
now i remembered that the stopwatch test is actually disabled on mac because it did not work even with higher epsilon

in general i'd prefer a test which instead of measuring time proves correctness by correct ordering, that would avoid problems with time measurement in test environment

so you can definitely replace this test with such a test of good ordering
so imagine two or more tasks (let's do two for simplicity, i'd do three for fun, but simplicity is better than fun) - A, B
and these guys want to cooperate to create a list of 100 integers from 1 to 100.

they will sync with this async event

so you create two events EA and EB, EA is set and EB is reset
you create two tasks over the same body and pass both events as the arguments, but for A, you pass events in order "EA, EB" and for B you pass events in order "EB, EA"

and you create shutdownEvent that is set to false

```
body(myEvent, theirEvent, shutdownEvent)
{
   loop until we are done
   {
       wait for myEvent + shutdownEvent
       reset myEvent
       get last number from the list
       add one to that number
       add the new number to the list
       if the newly added number is the limit to reach signal shutdownEvent
       set theirEvent
   }
}
```

the main thread should wait only for shutdownEvent and then asserts that the list is formed well

Aprogiena replied:
and we could also assert that when A wakes up then EB is reset and vice versa

noescape00 replied:
I'm really not sure if we should overcomplicate this. 
Ok, I agree that we should the <600 condition just in case. But the ordering test is overkill. 

>proves correctness by correct ordering

This wasn't the intention of this test. This test tests if the event was triggered after the 500ms delay and not before, that's all.

Aprogiena replied:
I know, but a little more complicated test would show that the async event is reliable. Maybe it is overkill and we can think of a much simpler test (although it is really 15 minutes of coding). 

The problem with this test is that it will possibly fail in some environments. The stopwatch tests that failed on Mac was more complex, so maybe this one would work, but conceptually, such measurement is not good and can fail (imagine your machine is under heavy stress, executing 200 threads and the thread that should jump of your task.delay won't get scheduled on time, the test will fail and it should not ...). So if you feel you can think out a simpler test, let's go for a simpler, but I'd like to avoid using stopwatch.

noescape00 replied:
Fixed. 

>so imagine two or more tasks (let's do two for simplicity, i'd do three for fun, but simplicity is better than fun)
Kept it simple. Had some fun. 
Now we have 10 tasks.

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+
+            CancellationTokenSource tokenSource = new CancellationTokenSource();
+
+            Task task = Task.Run(async () =>
```
this is meeeh
use `            CancellationTokenSource tokenSource = new CancellationTokenSource(500);` instead of this charade
and for the sake of all kittens, dispose your cancellation token sources!!!

noescape00 replied:
Ah, yeah. 
That was silly.

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -0,0 +1,145 @@
+        [Fact]
+        public async void AsyncManualResetEvent_CanBeCancelled()
+        {
+            AsyncManualResetEvent manualResetEvent = new AsyncManualResetEvent(false);
```
here and below for CTS we use `var`

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -127,5 +131,60 @@ public void AsyncManualResetEvent_MultipleWaitAsync_Set_IsCompleted()
+            var cts = new CancellationTokenSource();
+
+            var events = new List<AsyncManualResetEvent>();
+            for (int i = 0; i < tasksCount; ++i)
```
no `++i` in c#

Aprogiena commented on src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs:
```diff
@@ -127,5 +131,60 @@ public void AsyncManualResetEvent_MultipleWaitAsync_Set_IsCompleted()
+
+            var tasks = new List<Task>();
+
+            for (int i = 0; i < tasksCount; ++i)
```
no ++i

Aprogiena commented:
I'm happy with this now, tests seem to really test the class. So @noescape00 just pls check my last modifications if you are happy with them.

noescape00 commented:
@Aprogiena yes, looks good to me. 

>for (int i = 0; i < resultList.Count; i++)
>                Assert.Equal(i, resultList[i]);

This assertion loop looks unnecessary to me but I'm fine with it.
